nemo/collections/nlp/models/language_modeling/megatron_t5_model.py

```diff
@@ -0,0 +1,476 @@
+        eval_iters = (self.trainer.max_steps // self.trainer.val_check_interval + 1) * self.trainer.limit_val_batches
+        test_iters = self.trainer.limit_test_batches
+        train_valid_test_num_samples = [
+            self.trainer.max_steps * global_batch_size,
+            eval_iters * global_batch_size,
+            test_iters * global_batch_size,
+        ]
+        self._train_ds, self._validation_ds, self._test_ds = build_train_valid_test_datasets(
+            cfg=self.cfg,
+            trainer=self.trainer,
+            tokenizer=self.tokenizer,
+            data_prefix=self.cfg.data.data_prefix,
+            data_impl=self.cfg.data.data_impl,
+            splits_string=self.cfg.data.splits_string,
+            train_valid_test_num_samples=train_valid_test_num_samples,
+            max_seq_length=self.cfg.data.seq_length,
+            max_seq_length_dec=self.cfg.data.seq_length_dec,
+            masked_lm_prob=self.cfg.data.masked_lm_prob,
+            short_seq_prob=self.cfg.data.short_seq_prob,
+            seed=self.cfg.seed,
+            skip_warmup=self.cfg.data.skip_warmup,
+            dataset_type='t5',
+        )
+        logging.info(f'Length of train dataset: {len(self._train_ds)}')
+        logging.info(f'Length of val dataset: {len(self._validation_ds)}')
+        logging.info(f'Length of test dataset: {len(self._test_ds)}')
+        logging.info(f'Finished building T5 datasets.')
+        return self._train_ds, self._validation_ds, self._test_ds
+
+    def build_pretraining_data_loader(self, dataset, consumed_samples):
+        """Buld dataloader given an input dataset."""
+
+        if dataset is None:
+            return None
+
+        # Megatron sampler
+        if self.cfg.data.dataloader_type == 'single':
+            batch_sampler = MegatronPretrainingSampler(
+                total_samples=len(dataset),
+                consumed_samples=consumed_samples,
+                micro_batch_size=self.cfg.micro_batch_size,
+                data_parallel_rank=parallel_state.get_data_parallel_rank(),
+                data_parallel_size=parallel_state.get_data_parallel_world_size(),
+            )
+        elif self.cfg.data.dataloader_type == 'cyclic':
+            batch_sampler = MegatronPretrainingRandomSampler(
+                total_samples=len(dataset),
+                consumed_samples=consumed_samples,
+                micro_batch_size=self.cfg.micro_batch_size,
+                data_parallel_rank=parallel_state.get_data_parallel_rank(),
+                data_parallel_size=parallel_state.get_data_parallel_world_size(),
+            )
+        else:
+            raise Exception('{} dataloader type is not supported.'.format(self.cfg.dataloader_type))
+
+        # Torch dataloader.
+        return torch.utils.data.DataLoader(
+            dataset, batch_sampler=batch_sampler, num_workers=self.cfg.data.num_workers, pin_memory=True,
+        )
+
+    def setup(self, stage=None):
```

Comment:
Add a docstring explaining that this a PTL method and what it does.

Reply:
Done.